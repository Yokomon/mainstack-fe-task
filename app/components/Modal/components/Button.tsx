"use client";

import clsx from "clsx";
import React, { HTMLProps, ReactNode } from "react";

interface ButtonProps {
  children: ReactNode;
  className?: HTMLProps<HTMLElement>["className"];
  onClick?: () => void;
}

const Button: React.FC<ButtonProps> = ({ children, className, onClick }) => {
  return (
    <button
      onClick={onClick}
      className={clsx({
        ["flex shadow-sm text-sm font-medium tracking-tight items-center justify-center rounded-full px-12 ring-1 ring-gray-200 w-full p-3"]:
          true,
        [className!]: className,
      })}
    >
      {children}
    </button>
  );
};

export default Button;
