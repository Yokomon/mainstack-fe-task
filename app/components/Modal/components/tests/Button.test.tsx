import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import Button from "../Button";

describe("Modal Button Component", () => {
  it("renders the button with children", () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText("Click me")).toBeInTheDocument();
  });

  it("calls the onClick function when clicked", () => {
    const onClickMock = jest.fn();
    render(<Button onClick={onClickMock}>Click me</Button>);
    fireEvent.click(screen.getByText("Click me"));
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it("applies custom className if provided", () => {
    render(<Button className="custom-class">Click me</Button>);
    const button = screen.getByText("Click me");
    expect(button).toHaveClass("custom-class");
  });

  it("has default styles for basic rendering", () => {
    render(<Button>Click me</Button>);
    const button = screen.getByText("Click me");
    expect(button).toHaveClass(
      "flex",
      "shadow-sm",
      "text-sm",
      "items-center",
      "justify-center",
      "rounded-full",
      "px-12",
      "ring-1",
      "ring-gray-200",
      "w-full",
      "p-3"
    );
  });

  it("applies additional styles when className is provided", () => {
    render(<Button className="custom-class">Click me</Button>);
    const button = screen.getByText("Click me");
    expect(button).toHaveClass(
      "flex",
      "shadow-sm",
      "text-sm",
      "items-center",
      "justify-center",
      "rounded-full",
      "px-12",
      "ring-1",
      "ring-gray-200",
      "w-full",
      "p-3",
      "custom-class"
    );
  });

  it("renders with a custom onClick function", () => {
    const onClickMock = jest.fn();
    render(<Button onClick={onClickMock}>Click me</Button>);
    fireEvent.click(screen.getByText("Click me"));
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it("renders with custom children", () => {
    render(<Button>ðŸš€ Launch</Button>);
    expect(screen.getByText("ðŸš€ Launch")).toBeInTheDocument();
  });

  it("renders without crashing", () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText("Click me")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Button>Click me</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});
