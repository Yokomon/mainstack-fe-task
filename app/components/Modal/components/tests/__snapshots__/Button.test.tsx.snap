// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Button Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="flex shadow-sm text-sm font-medium tracking-tight items-center justify-center rounded-full px-12 ring-1 ring-gray-200 w-full p-3"
  >
    Click me
  </button>
</DocumentFragment>
`;
